dynamiqs/mesolve/rouchon.py

```diff
@@ -24,49 +24,36 @@ class MERouchon(MESolver, AdjointFixedSolver):
 
 
 class MERouchon1(MERouchon):
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-
-        # forward time operators
-        # M0: (b_H, 1, n, n)
-        # M1s: (1, len(L), n, n)
-        self.M0 = cache(lambda Hnh: self.I - 1j * self.dt * Hnh)
-        self.M1s = sqrt(self.dt) * self.L
-
-        # reverse time operators
-        # M0rev: (b_H, 1, n, n)
-        # M1srev: (1, len(L), n, n)
-        self.M0rev = cache(lambda Hnh: self.I + 1j * self.dt * Hnh)
-        self.M1srev = self.M1s
-
-        if self.options.normalize:
-            # `R` is close to identity but not exactly, we inverse it to normalize the
-            # Kraus map in order to have a trace-preserving scheme
-            self.R = cache(lambda M0: M0.mH @ M0 + self.dt * self.sum_LdagL)
-            self.Rrev = cache(lambda M0rev: M0rev.mH @ M0rev - self.dt * self.sum_LdagL)
-
-            if self.H.is_constant:
-                # if H is constant, we normalize the operators once by computing
-                # `S = sqrt(R)^-1` and applying is to the `Ms` operators
-                H = self.H(0.0)
-                Hnh = self.Hnh(H)
-                M0 = self.M0(Hnh)
-                M0rev = self.M0rev(Hnh)
-
-                S = inv_sqrtm(self.R(M0))
-                Srev = inv_sqrtm(self.Rrev(M0rev))
-
-                self.M0 = cache(lambda Hnh: M0 @ S)
-                self.M1s = self.M1s @ S
-                self.M0rev = cache(lambda Hnh: Srev @ M0rev)
-                self.M1srev = Srev @ self.M1s
-            else:
-                # if `H` is time-dependent, we normalize the map at each time step by
-                # inverting `R` using its Cholesky decomposition `R = T @ T.mT`
-                self.T = cache(lambda R: cholesky(R)[0])  # lower triangular
-                self.Trev = cache(lambda Rrev: cholesky(Rrev)[0])  # lower triangular
-
-        self.cholesky_normalization = self.options.normalize and not self.H.is_constant
+    @cache(maxsize=2)
+    def R(self, M0: Tensor, dt: float) -> Tensor:
+        # `R` is close to identity but not exactly, we inverse it to normalize the
+        # Kraus map in order to have a trace-preserving scheme
+        # -> (b_H, 1, n, n)
+        return M0.mH @ M0 + dt * self.sum_LdagL
+
+    @cache(maxsize=2)
+    def Ms(self, Hnh: Tensor, dt: float) -> tuple(Tensor, Tensor):
+        # Kraus operators
+        # -> (b_H, 1, n, n), (1, len(L), n, n)
+        M0 = self.I - 1j * dt * Hnh  # (b_H, 1, n, n)
+        M1s = sqrt(self.dt) * self.L  # (1, len(L), n, n)
```

Comment:
You use `selt.dt` here but `dt` is a method argument. Throughout the class methods (also in e.g. `forward`), can you normalize the use of `dt` and `self.dt`? Use one or the other every time. This would make more sense IMO.